lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+/*
+Copyright 2024 Matteo Cenzato
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include <utility>
+#include <iostream>
+#include <limits>
+#include <optional>
+#include <string>
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/DenseAnalysis.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/Interfaces/ControlFlowInterfaces.h"
+#include "rlc/dialect/Operations.hpp"
+#include "rlc/dialect/Passes.hpp"
+#include "rlc/dialect/conversion/TypeConverter.h"
+
+// Add the map key to the llvm scope
+// I have to admit that if this works the llms will take over the world someday
+// Ok it compiles
+// It seems to work. Wow
+namespace llvm{
+	template <>
+	struct llvm::DenseMapInfo<std::pair<mlir::Value,bool>>{
+		static std::pair<mlir::Value,bool> getEmptyKey(){
+			return std::make_pair(llvm::DenseMapInfo<mlir::Value>::getEmptyKey(),false);
+		}
+		static std::pair<mlir::Value, bool> getTombstoneKey() {
+			return std::make_pair(DenseMapInfo<mlir::Value>::getTombstoneKey(), false);
+		}
+		static unsigned getHashValue(const std::pair<mlir::Value, bool> &Val) {
+			return hash_combine(DenseMapInfo<mlir::Value>::getHashValue(Val.first), Val.second);
+		}
+		static bool isEqual(const std::pair<mlir::Value, bool> &lhs, const std::pair<mlir::Value, bool> &rhs) {
+			return DenseMapInfo<mlir::Value>::isEqual(lhs.first, rhs.first) && lhs.second == rhs.second;
+		}
+	};
+}
+
+namespace mlir::rlc
+{
+	namespace
+	{
+		// Represents the pair (min, max) of values that can be assigned to a variable 
+		// in a function
+
+		// We need an abstraction to represent the range, because we need to 
+		// define our particular bounds. 
+		// Indeed we require an uninitialized state (we call it BOTTOM) and a TOP
+		// NB: TOP is the maximum practical value the variable can assume -> for the lhs is INT_MIN and rhs is INT_MAX
+		// TODO: what to do when we finish in an invalid range? E.G. (1,-1)
+		
+		// TODO: I could think about implementing disjointed ranges but the algorithms starts to become 
+		// very complicated (think about a multiplication, definitely not trivial)
+		class IntegerRange{
```

Comment:
can we replace this class with https://llvm.org/doxygen/classllvm_1_1ConstantRange.html ?

i think the other time we talked about going from a pair of optional ints to a options pair of int, if that is still somewhere we want to do, then llvm::ConstantRange should be identical to this one

Reply:
yes we can, it will take some time but it should definitely be possible (now the pair of optionals is almost useless since the nullptr is treated as a TOP value when joining).

Reply:
ok, then it is worth doing